application-task-default/src/main/java/com/xwiki/task/internal/TaskObjectUpdateEventListener.java

```diff
@@ -105,37 +118,68 @@ protected void processEvent(XWikiDocument document, XWikiContext context, Event
                         String.format("Task [%s] has been updated!", taskObj.getDocumentReference()), context);
                 }
             }
-            context.put(TASK_UPDATE_FLAG, null);
         } catch (XWikiException e) {
             logger.warn("Failed to process the owner document of the task [{}]: [{}].", taskOwnerRef,
                 ExceptionUtils.getRootCauseMessage(e));
+        } finally {
+            context.put(TASK_UPDATE_FLAG, null);
         }
     }
 
     private boolean handleDeleteEvent(XWikiDocument document, XWikiContext context, Event event)
     {
         if (event instanceof DocumentDeletingEvent) {
+            if (context.get(TASK_UPDATE_FLAG) != null) {
+                return true;
+            }
             try {
+                // We can't use the doc provided to the listener because it does not contain the xwiki objects.
                 XWikiDocument actualDoc = context.getWiki().getDocument(document.getDocumentReference(), context);
                 BaseObject object = actualDoc.getXObject(TASK_CLASS_REFERENCE);
-                if (object != null && !object.getStringValue(Task.OWNER).isEmpty()) {
-                    XWikiDocument hostDocument = context.getWiki().getDocument(
-                        resolver.resolve(object.getStringValue(Task.OWNER), document.getDocumentReference()), context);
-                    hostDocument.setContent(taskXDOMProcessor.removeTaskMacroCall(document.getDocumentReference(),
-                        hostDocument.getDocumentReference(), hostDocument.getXDOM(), hostDocument.getSyntax()));
-                    context.getWiki().saveDocument(hostDocument,
-                        String.format("Removed the task with the reference of [%s]", document.getDocumentReference()),
-                        context);
+                if (object == null || object.getStringValue(Task.OWNER).isEmpty()) {
+                    return true;
+                }
+                DocumentReference hostDocumentReference = resolver.resolve(object.getStringValue(Task.OWNER),
+                    document.getDocumentReference());
+                // If we are inside a refactoring job that also deletes the owner, no need to update the macro call.
+                List<String> deleteJobGroup = new ArrayList<>();
+                deleteJobGroup.add("refactoring");
+                for (EntityReference entityReference : document.getDocumentReference().getLastSpaceReference()
+                    .getReversedReferenceChain()) {
+                    deleteJobGroup.add(entityReference.getName());
                 }
+                Job deletingJob = executor.getCurrentJob(new JobGroupPath(deleteJobGroup));
```

Comment:
Is this an exact match on the job group? What if the refactoring job was triggered by deleting some ancestor page that is a few levels above the task page?